lib/ipam/ipam.go

```diff
@@ -119,7 +125,24 @@ func (c ipamClient) AutoAssign(ctx context.Context, args AutoAssignArgs) ([]net.
 		v6list, err = c.autoAssign(ctx, args.Num6, args.HandleID, args.Attrs, args.IPv6Pools, 6, hostname, args.MaxBlocksPerHost, args.HostReservedAttrIPv6s)
 		if err != nil {
 			log.Errorf("Error assigning IPV6 addresses: %v", err)
-			return v4list, v6list, err
+		}
+
+		if len(v6list) != args.Num6 {
```

Comment:
I think this fundamentally changes the way this function was intended to behave. Admittedly, we're not really using it in that way ourselves, but the intention of the function was always that it would allocate as many addresses as it could in order to satisfy the request and return them to the caller.

Imagine this scenario:
- Call auto-assign asking for 3 IPv6 addresses and 1 IPv4 address.
- We only allocate 2 IPv6 addresses, and 1 IPv4 address.

Currently, the function will return 2 IPv6 addresses and 1 IPv4 address.

With this change, it will return 2 IPv6 addresses and 0 IPv4 addresses.

Reply:
Same goes for the reverse scenario (where we hit the v6 case above). We'll no longer be doing a best-effort allocation of as many addresses as we can to satisfy the request.